widget/editor.go

```diff
@@ -648,36 +826,56 @@ func (e *Editor) invalidate() {
 
 // Delete runes from the caret position. The sign of runes specifies the
 // direction to delete: positive is forward, negative is backward.
+//
+// If there is a selection, it is deleted and counts as a single rune.
 func (e *Editor) Delete(runes int) {
-	e.rr.deleteRunes(runes)
-	e.caret.xoff = 0
+	if runes == 0 {
+		return
+	}
+
+	if l := e.selStart.ofs - e.caret.pos.ofs; l != 0 {
+		e.caret.pos.ofs = e.rr.deleteRunes(e.caret.pos.ofs, l)
+		runes -= sign(runes)
+	}
+
+	e.caret.pos.ofs = e.rr.deleteRunes(e.caret.pos.ofs, runes)
+	e.caret.pos.xoff = 0
+	e.ClearSelection()
 	e.invalidate()
 }
 
-// Insert inserts text at the caret, moving the caret forward.
+// Insert inserts text at the caret, moving the caret forward. If there is a
+// selection, Insert overwrites it.
 func (e *Editor) Insert(s string) {
 	e.append(s)
 	e.caret.scroll = true
-	e.invalidate()
 }
 
+// append inserts s at the cursor, leaving the caret is at the end of s. If
+// there is a selection, append overwrites it.
+// xxx|yyy + append zzz => xxxzzz|yyy
 func (e *Editor) append(s string) {
 	e.prepend(s)
-	e.rr.caret += len(s)
+	e.caret.pos.ofs += len(s)
+	e.selStart.ofs = e.caret.pos.ofs
 }
 
+// prepend inserts s after the cursor; the caret does not change. If there is
+// a selection, prepend overwrites it.
+// xxx|yyy + prepend zzz => xxx|zzzyyy
 func (e *Editor) prepend(s string) {
 	if e.SingleLine {
 		s = strings.ReplaceAll(s, "\n", " ")
 	}
-	e.rr.prepend(s)
-	e.caret.xoff = 0
+	e.caret.pos.ofs = e.rr.deleteRunes(e.caret.pos.ofs, e.selStart.ofs-e.caret.pos.ofs) // Delete any selection first.
+	e.rr.prepend(e.caret.pos.ofs, s)
+	e.caret.pos.xoff = 0
 	e.invalidate()
 }
 
-func (e *Editor) movePages(pages int) {
+func (e *Editor) movePages(pages int, selAct selectionAction) {
```

Comment:
It's weird to have selectionAction in unexported API, and (start, end) offsets in exported API (MoveCaret). In fact, I find it weird to have different methods for similar actions (moving the caret). I suggest to replace them all with a single method:

```
type MoveUnit

const (
    MoveRune MoveUnit = iota
    MoveWord
    MovePage
    // Maybe?
    MoveLineStart
    MoveLineEnd
    // If you'd like to replace SetCaret as well.
    MoveAbsolute
)

// MoveCaret move the start and end of the caret start and end
// units, respectively.
func (e *Editor) MoveCaret(unit MoveUnit, start, end int)
```

With something like MoveAbsolute you could even get rid of SetCaret.

MoveUnit is not the greatest name, I know, and it's ok to postpone this API discussion for later.

Reply:
> it's ok to postpone this API discussion for later

Cool, then I will.  😄